include/kamping/p2p/sendrecv.hpp

```diff
@@ -0,0 +1,267 @@
+        Args,
+        KAMPING_REQUIRED_PARAMETERS(send_buf, destination),
+        KAMPING_OPTIONAL_PARAMETERS(
+            send_count,
+            send_type,
+            send_tag,
+            recv_buf,
+            recv_tag,
+            source,
+            recv_type,
+            status,
+            recv_count
+        )
+    );
+
+    auto send_buf = internal::select_parameter_type<internal::ParameterType::send_buf>(args...)
+                        .template construct_buffer_or_rebind<UnusedRebindContainer, serialization_support_tag>();
+
+    using default_send_count_type = decltype(kamping::send_count_out());
+    auto send_count =
+        internal::select_parameter_type_or_default<internal::ParameterType::send_count, default_send_count_type>(
+            {},
+            args...
+        )
+            .construct_buffer_or_rebind();
+    if constexpr (has_to_be_computed<decltype(send_count)>) {
+        send_count.underlying() = asserting_cast<int>(send_buf.size());
+    }
+
+    using send_value_type = typename std::remove_reference_t<decltype(send_buf)>::value_type;
+    auto send_type        = internal::determine_mpi_send_datatype<send_value_type>(args...);
+
+    auto const&    destination = internal::select_parameter_type<internal::ParameterType::destination>(args...);
+    constexpr auto rank_type   = std::remove_reference_t<decltype(destination)>::rank_type;
+    static_assert(
+        rank_type == RankType::value || rank_type == RankType::null,
+        "Please provide an explicit destination or destination(ranks::null)."
+    );
+
+    using default_send_tag_type = decltype(kamping::tag(this->default_tag()));
+    auto&& send_tag_param =
+        internal::select_parameter_type_or_default<internal::ParameterType::send_tag, default_send_tag_type>(
+            std::tuple(this->default_tag()),
+            args...
+        );
+
+    // this ensures that the user does not try to pass MPI_ANY_TAG, which is not allowed for the send tag
+    static_assert(
+        std::remove_reference_t<decltype(send_tag_param)>::tag_type == TagType::value,
+        "Please provide a send tag for the message."
+    );
+    int send_tag = send_tag_param.tag();
+    KASSERT(
+        Environment<>::is_valid_tag(send_tag),
+        "invalid send tag " << send_tag << ", must be in range [0, " << Environment<>::tag_upper_bound() << "]"
+    );
+
+    using default_recv_buf_type = decltype(kamping::recv_buf(alloc_new<DefaultContainerType<recv_value_type_tparam>>));
+    auto recv_buf =
+        internal::select_parameter_type_or_default<internal::ParameterType::recv_buf, default_recv_buf_type>(
+            std::tuple(),
+            args...
+        )
+            .template construct_buffer_or_rebind<DefaultContainerType, internal::serialization_support_tag>();
+
+    KASSERT(
+        !(send_buf.size() == recv_buf.size() && send_buf.data() == recv_buf.data()),
+        "Send buffer and recv buffer can not be the same."
+    );
```

Comment:
Is this a required by the standard? Also, I am not really source about using the buffer's size here, because it could also have been provided explicitely by the user via send_count. Maybe leave this check out alltogether?